packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -15,71 +16,67 @@ import "@opengsn/contracts/src/utils/GsnUtils.sol";
 import "./interfaces/IChainlinkOracle.sol";
 
 import "./helpers/UniswapV3Helper.sol";
-import "./helpers/PermitInterfaceDAI.sol";
+import "../../contracts/src/interfaces/IERC20Token.sol";
 
 /**
  * A paymaster allowing addresses holding ERC20 tokens with 'permit' functionality
  * to pay for a GSN transaction.
  */
 contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
 
+    using SafeERC20 for IERC20;
+
     event Received(address indexed sender, uint256 eth);
-    event TokensCharged(uint256 gasUseWithoutPost, uint256 gasJustPost, uint256 ethActualCharge);
+    event FundingNeeded();
+    event TokensCharged(uint256 gasUseWithoutPost, uint256 gasJustPost, uint256 tokenActualCharge, uint256 ethActualCharge);
 
-    IChainlinkOracle public priceFeed;
+    mapping(IERC20Metadata => IChainlinkOracle) public priceFeeds;
```

Comment:
its bad to have multiple mappings with same key.
seems better to have a single mapping to a struct (better both in code readability and performance.)
along the way, it allows packing values together: (mapping address=>uint24) takes full uint256 for each member, but struct { uint24, address, } are packed into a single storage cell (and the mapping key is indexed once)

Reply:
Yes, I added them one by one and forgot about it. I'll pack them to a struct